### ISSUE TITLE
Death Grip. The pull effect shouldn't work on (raid) bosses. Also, the spell script code is strange.

### ISSUE THREAD

**Issue by chaodhib** [TrinityCore/TrinityCore#14870]
First of all, the pull effect of Death Grip shouldn't affect (raid) bosses while the taunt should work as a general rule (it specifically depends on the boss. some bosses are tauntable. others not).

Second of all, to me the way the spell is scripted is very strange. Dks have access to the Death Grip spell 49576. That spell has 3 effects:

1. Effect 0: SPELL_EFFECT_DUMMY (TARGET_UNIT_TARGET_ENEMY, TARGET_DEST_CASTER_FRONT) with Effect Mechanic=MECHANIC_GRIP.
2. Effect 1: SPELL_EFFECT_APPLY_AURA. It applies Taunt aura on target
3. Effect 2: SPELL_EFFECT_ATTACK_ME.

The weird part is the way the dummy part of Effect 0 is scripted (https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_dk.cpp#L1648). After several checks are done, it calls the spell 49560 which doesnt exist the in DBC but it has an entry in spell_script_name which link it to another spell script (https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_dk.cpp#L642) where the job is actually done. Why this AFAIK useless spell and second spell script when the content of the second spell script could be moved into the first spell script?

CORE revision number: d86ff08 (branch 3.3.5a, 15/06/10)
TDB 335.58 +updates up until 2015_06_08_00_world.

Thank you in advance for your input.
 Chaodhib

**Comment by Shauren** [comment]
49560 doesn't exist? Give me a break and at least look at dbc files... effects 1 & 2 you listed are actually on 49560

**Comment by chaodhib** [comment]
@Shauren You are correct actually. I messed up. My bad. 

First point holds still true though.

**Comment by P-Kito** [comment]
Huh. For me it does not work for raid bosses Oo

**Comment by Killyana** [comment]
We know how the death grip should works with bosses also the 2 effects are handled by mechanic_immune_mask (32	0x0000 0020	MECHANIC_GRIP) and the extrflag (256	CREATURE_FLAG_EXTRA_NO_TAUNT).
So what's the issue here?

**Comment by Goatform** [comment]
Adding functions such as isWorldBoss and isElite and using return; in spell script should fix this issue.

**Comment by chaodhib** [comment]
@Killyana @P-Kito I tried death gripping Gluth and Patchwork and in both cases, it worked. That's the issue.

@Goatform yeah. Remplacing https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_dk.cpp#L658 by:
```C++
if (!target->HasAuraType(SPELL_AURA_DEFLECT_SPELLS) && !target->IsDungeonBoss() &&!target->isWorldBoss())
```
(and changing https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_dk.cpp#L654 int32 damage to int32 spellID because the name is confusing)

would solve the issue I think. Not sure if that's the best way to handle this though. First of all, there would be no "immune" to the pull component message sent to the player. In retail, there are (at best) 2 immune msgs displayed: one "Immune" if the target is immune to taunt (example: a player). and another "immune" if the target is immune to the pull effect (example: a boss).

**Comment by Killyana** [comment]
I'm not sure if it's a good idea to hardcore the immunity (maybe some bosses are not immune to the gripping effect) you can just add 32 to the "mechanic_immune_mask" on creature_template.

**Comment by Goatform** [comment]
I think that all bosses are immune on that pull effect.

**Comment by chaodhib** [comment]
I never played dk in wotlk retail so I don't have first hand experience and I might be wrong but I think Goatform is right.

**Comment by Shauren** [comment]
No, this is not the case for dungeon bosses. There are a few exceptions that are grippable (using current expansion as example, http://www.wowhead.com/npc=86243 )

**Comment by chaodhib** [comment]
So I tested Death Grip on most bosses of Naxx and it worked on each one. Feels very strange to pull something 20 times bigger than yourself :x

Anyway, do we have any sniff on Naxx's bosses? And what informations are given by sniffs? does it give the "mechanic_immune_mask" taken from DB or does it give every immunity, including the ones added at run time?

**Comment by Goatform** [comment]
I still think that it's much better to fix it in SpellScript instead of doing that.

**Comment by Aokromes** [comment]
@chaodhib we have naxx sniffs, but sniffs don't contains something similar to mechanic_immune_mask, only testing with chars having the desired mechanic_immune_mask is possible to dig sniffs searching if they are immune to it.

**Comment by chaodhib** [comment]
So, how to make bosses (or any creature) immune to (only the pull component of) death grip?

So far, the only way found is to hard code it somewhere (either in the spell script or make the creature immune to the specific spell using its spell ID). I think we can all agree it's ugly and it will most likely have drawbacks later for some bosses scripts of some encounter.

Ideally, I think we want to be able to make a creature immune to the pull effect of the death grip using the immunity to MECHANIC_GRIP. Problem is, the spell itself has MECHANIC field set to MECHANIC_GRIP.

Is it possible that in retail servers, the Mechanic field is not used when determining any applicable immunity and instead, only the 3 Effect Mechanic fields are used?

**Comment by chaodhib** [comment]
Concrete example of what should be possible if this issue is fixed (taken from a retail 4.X video of the IC encounter https://www.youtube.com/watch?v=7cALSvQFoFI):
![vlcsnap-2015-08-25-20h43m04s082](https://cloud.githubusercontent.com/assets/6612710/9475988/7005bac4-4b6a-11e5-8579-7d07d42c04d8.png)
Notice: 
* the debuff on Steelbreaker. it's the 3s taunt debuff.
* the "immune" message referring to the pull effect of the spell that has failed.

On the other hand, later in the same video, in the same encounter, we can see Stormcaller Brundir being affected by both effects (pull & taunt).

In definitive, we should have bosses that are either:
1. Immune to the pull effect. Not immune to the taunt (ex Steelbreaker. Most woltk bosses are in this category I believe)
2. Non immune to both effects. (ex Brundir)
3. immune to both (Gluth after he hard enrages for example)

and possibly:
4. immune to taunt but not the pull effect

There is not a single solution in this thread so far that allows us to put a creature in each of these 3 states.
* Setting a creature immune to MECHANIC_GRIP spells (in the db or by code) result in an immunity to both the pull & the taunt. State 1 impossible.
* ```me->ApplySpellImmune(0, IMMUNITY_ID, 49560, true)``` currently used in some scripts (Vezzax, Freya, possibly others) makes the boss immune to the entire spell -> State 1 impossible.
* Adding a condition on the spell script (such as, dungeon & raid bosses should be immune to the pull effect) is bad too since we wont be able to have exceptions. State 2 wont be possible
* Even "ugly solutions" such as ```me->ApplySpellImmune(0, IMMUNITY_EFFECT, SPELL_EFFECT_JUMP_DEST, true);``` and ```me->ApplySpellImmune(0, IMMUNITY_ID, 49575, true); // jump aspect of Death Grip only``` dont work because the spell 49575 has Trigered flag set to true.

I tested all of this. 

In definitive, it would be great if an experienced dev could take a look at this issue. Also, creatures have an extra flag for taunt. Maybe it is time we have an extra flag for Death Grip's pull effect?

**Comment by Aokromes** [comment]
still valid?

**Comment by chaodhib** [comment]
yes

**Comment by CraftedRO** [comment]
8b1d6f91ad58fd31d2c5f203d11ff251689c8aed looks like the given example is no longer valid:

https://github.com/user-attachments/assets/2a23343a-a335-4d22-b10d-bd702e5eac01
